#include "MainWindow.h"
#include "ui_MainWindow.h"
#include "widgets/WarehouseWidget.h"
#include "widgets/ReportsWidget.h"
#include "widgets/RequestsWidget.h"
#include "gui/RegistrationDialog.h"
#include "database/DatabaseManager.h"

#include <QStatusBar>
#include <QMenuBar>
#include <QToolBar>
#include <QMessageBox>
#include <QDateTime>
#include <QIcon>

MainWindow::MainWindow(const User& user, QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
    , m_currentUser(user)
    , m_warehouseWidget(nullptr)
    , m_reportsWidget(nullptr)
    , m_requestsWidget(nullptr)
{
    ui->setupUi(this);
    
    setWindowTitle("–°–∫–ª–∞–¥—Å–∫–∞—è —Å–∏—Å—Ç–µ–º–∞ HK - " + user.fullName());
    
    setupUi();
    setupMenuBar();
    setupToolBar();
    setupConnections();
    applyUserPermissions();
    
    statusBar()->showMessage(QString("–ü–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å: %1 (%2) | %3")
        .arg(user.fullName())
        .arg(user.role())
        .arg(QDateTime::currentDateTime().toString("dd.MM.yyyy HH:mm:ss")));
}

MainWindow::~MainWindow()
{
    delete ui;
}

void MainWindow::setupUi()
{
    // –°–æ–∑–¥–∞–µ–º –≤–∏–¥–∂–µ—Ç—ã
    m_warehouseWidget = new WarehouseWidget(m_currentUser, this);
    m_reportsWidget = new ReportsWidget(m_currentUser, this);
    m_requestsWidget = new RequestsWidget(m_currentUser, this);
    
    // –î–æ–±–∞–≤–ª—è–µ–º –≤–∏–¥–∂–µ—Ç—ã –≤ stacked widget
    ui->stackedWidget->addWidget(m_warehouseWidget);
    ui->stackedWidget->addWidget(m_reportsWidget);
    ui->stackedWidget->addWidget(m_requestsWidget);
    
    // –ü–æ–∫–∞–∑—ã–≤–∞–µ–º –≤–∏–¥–∂–µ—Ç —Å–∫–ª–∞–¥–∞ –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é
    ui->stackedWidget->setCurrentWidget(m_warehouseWidget);
    
    // –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –∏–∫–æ–Ω–∫—É
    setWindowIcon(QIcon(":/icons/warehouse.png"));
}

void MainWindow::setupMenuBar()
{
    // –ú–µ–Ω—é –§–∞–π–ª
    QMenu *fileMenu = menuBar()->addMenu("&–§–∞–π–ª");
    
    QAction *exportAction = new QAction("–≠–∫—Å–ø–æ—Ä—Ç –¥–∞–Ω–Ω—ã—Ö", this);
    exportAction->setShortcut(QKeySequence::Save);
    connect(exportAction, &QAction::triggered, this, [this]() {
        m_warehouseWidget->onExportDataClicked();
    });
    fileMenu->addAction(exportAction);
    
    fileMenu->addSeparator();
    
    QAction *exitAction = new QAction("–í—ã—Ö–æ–¥", this);
    exitAction->setShortcut(QKeySequence::Quit);
    connect(exitAction, &QAction::triggered, this, &MainWindow::onExitButtonClicked);
    fileMenu->addAction(exitAction);
    
    // –ú–µ–Ω—é –ü—Ä–∞–≤–∫–∞
    QMenu *editMenu = menuBar()->addMenu("&–ü—Ä–∞–≤–∫–∞");
    
    QAction *addProductAction = new QAction("–î–æ–±–∞–≤–∏—Ç—å —Ç–æ–≤–∞—Ä", this);
    addProductAction->setShortcut(QKeySequence::New);
    connect(addProductAction, &QAction::triggered, m_warehouseWidget, &WarehouseWidget::onAddProductClicked);
    editMenu->addAction(addProductAction);
    
    QAction *refreshAction = new QAction("–û–±–Ω–æ–≤–∏—Ç—å", this);
    refreshAction->setShortcut(QKeySequence::Refresh);
    connect(refreshAction, &QAction::triggered, this, [this]() {
        QWidget *current = ui->stackedWidget->currentWidget();
        if (current == m_warehouseWidget) {
            m_warehouseWidget->refreshData();
        } else if (current == m_reportsWidget) {
            m_reportsWidget->refresh();
        } else if (current == m_requestsWidget) {
            m_requestsWidget->refresh();
        }
    });
    editMenu->addAction(refreshAction);
    
    // –ú–µ–Ω—é –í–∏–¥
    QMenu *viewMenu = menuBar()->addMenu("&–í–∏–¥");
    
    QAction *warehouseAction = new QAction("–°–∫–ª–∞–¥", this);
    connect(warehouseAction, &QAction::triggered, this, &MainWindow::onWarehouseButtonClicked);
    viewMenu->addAction(warehouseAction);
    
    QAction *requestsAction = new QAction("–ó–∞—è–≤–∫–∏", this);
    connect(requestsAction, &QAction::triggered, this, &MainWindow::onRequestsButtonClicked);
    viewMenu->addAction(requestsAction);
    
    QAction *reportsAction = new QAction("–û—Ç—á–µ—Ç—ã", this);
    connect(reportsAction, &QAction::triggered, this, &MainWindow::onReportsButtonClicked);
    viewMenu->addAction(reportsAction);
    
    // –ú–µ–Ω—é –ü–æ–º–æ—â—å
    QMenu *helpMenu = menuBar()->addMenu("&–ü–æ–º–æ—â—å");
    
    QAction *aboutAction = new QAction("–û –ø—Ä–æ–≥—Ä–∞–º–º–µ", this);
    connect(aboutAction, &QAction::triggered, this, [this]() {
        QMessageBox::about(this, "–û –ø—Ä–æ–≥—Ä–∞–º–º–µ",
            "<h3>–°–∫–ª–∞–¥—Å–∫–∞—è —Å–∏—Å—Ç–µ–º–∞ HK v2.0</h3>"
            "<p>–°–∏—Å—Ç–µ–º–∞ —É–ø—Ä–∞–≤–ª–µ–Ω–∏—è —Å–∫–ª–∞–¥–æ–º —Å –ø–æ–¥–¥–µ—Ä–∂–∫–æ–π QR-–∫–æ–¥–æ–≤</p>"
            "<p>–†–∞–∑—Ä–∞–±–æ—Ç—á–∏–∫: WarehouseHK Team</p>"
            "<p>¬© 2024 –í—Å–µ –ø—Ä–∞–≤–∞ –∑–∞—â–∏—â–µ–Ω—ã</p>");
    });
    helpMenu->addAction(aboutAction);
}

void MainWindow::setupToolBar()
{
    QToolBar *mainToolBar = addToolBar("–û—Å–Ω–æ–≤–Ω—ã–µ —Ñ—É–Ω–∫—Ü–∏–∏");
    mainToolBar->setMovable(false);
    
    // –ö–Ω–æ–ø–∫–∞ –°–∫–ª–∞–¥
    QAction *warehouseAction = new QAction("üì¶ –°–∫–ª–∞–¥", this);
    warehouseAction->setToolTip("–£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ —Ç–æ–≤–∞—Ä–∞–º–∏ –∏ –∫–æ–Ω—Ç–µ–π–Ω–µ—Ä–∞–º–∏");
    connect(warehouseAction, &QAction::triggered, this, &MainWindow::onWarehouseButtonClicked);
    mainToolBar->addAction(warehouseAction);
    
    // –ö–Ω–æ–ø–∫–∞ –ó–∞—è–≤–∫–∏
    QAction *requestsAction = new QAction("üìã –ó–∞—è–≤–∫–∏", this);
    requestsAction->setToolTip("–£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ –∑–∞—è–≤–∫–∞–º–∏ –Ω–∞ –ø–µ—Ä–µ–º–µ—â–µ–Ω–∏–µ");
    connect(requestsAction, &QAction::triggered, this, &MainWindow::onRequestsButtonClicked);
    mainToolBar->addAction(requestsAction);
    
    // –ö–Ω–æ–ø–∫–∞ –û—Ç—á–µ—Ç—ã
    QAction *reportsAction = new QAction("üìä –û—Ç—á–µ—Ç—ã", this);
    reportsAction->setToolTip("–ü—Ä–æ—Å–º–æ—Ç—Ä –æ—Ç—á–µ—Ç–æ–≤ –∏ –∞–Ω–∞–ª–∏—Ç–∏–∫–∞");
    connect(reportsAction, &QAction::triggered, this, &MainWindow::onReportsButtonClicked);
    mainToolBar->addAction(reportsAction);
    
    mainToolBar->addSeparator();
    
    // –ö–Ω–æ–ø–∫–∞ –ü–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª–∏ (—Ç–æ–ª—å–∫–æ –¥–ª—è –∞–¥–º–∏–Ω–∞)
    if (m_currentUser.isAdmin()) {
        QAction *usersAction = new QAction("üë• –ü–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª–∏", this);
        usersAction->setToolTip("–£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è–º–∏ —Å–∏—Å—Ç–µ–º—ã");
        connect(usersAction, &QAction::triggered, this, &MainWindow::onUsersButtonClicked);
        mainToolBar->addAction(usersAction);
    }
    
    mainToolBar->addSeparator();
    
    // –ö–Ω–æ–ø–∫–∞ –ù–∞—Å—Ç—Ä–æ–π–∫–∏
    QAction *settingsAction = new QAction("‚öôÔ∏è –ù–∞—Å—Ç—Ä–æ–π–∫–∏", this);
    connect(settingsAction, &QAction::triggered, this, &MainWindow::onSettingsButtonClicked);
    mainToolBar->addAction(settingsAction);
    
    // –ö–Ω–æ–ø–∫–∞ –í—ã—Ö–æ–¥
    QAction *exitAction = new QAction("üö™ –í—ã—Ö–æ–¥", this);
    exitAction->setToolTip("–í—ã–π—Ç–∏ –∏–∑ —Å–∏—Å—Ç–µ–º—ã");
    connect(exitAction, &QAction::triggered, this, &MainWindow::onExitButtonClicked);
    mainToolBar->addAction(exitAction);
}

void MainWindow::setupConnections()
{
    connect(ui->warehouseButton, &QPushButton::clicked, this, &MainWindow::onWarehouseButtonClicked);
    connect(ui->reportsButton, &QPushButton::clicked, this, &MainWindow::onReportsButtonClicked);
    connect(ui->requestsButton, &QPushButton::clicked, this, &MainWindow::onRequestsButtonClicked);
    connect(ui->usersButton, &QPushButton::clicked, this, &MainWindow::onUsersButtonClicked);
    connect(ui->settingsButton, &QPushButton::clicked, this, &MainWindow::onSettingsButtonClicked);
    connect(ui->exitButton, &QPushButton::clicked, this, &MainWindow::onExitButtonClicked);
}

void MainWindow::applyUserPermissions()
{
    QString role = m_currentUser.role();
    
    // –ü—Ä–∏–º–µ–Ω—è–µ–º –æ–≥—Ä–∞–Ω–∏—á–µ–Ω–∏—è
    if (role == "worker") {
        ui->usersButton->setEnabled(false);
        QList<QAction*> actions = findChildren<QAction*>();
        for (QAction* action : actions) {
            if (action->text().contains("–ü–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª–∏")) {
                action->setEnabled(false);
            }
        }
    }
}

void MainWindow::onWarehouseButtonClicked()
{
    ui->stackedWidget->setCurrentWidget(m_warehouseWidget);
    statusBar()->showMessage("–†–µ–∂–∏–º: –£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ —Å–∫–ª–∞–¥–æ–º");
}

void MainWindow::onReportsButtonClicked()
{
    ui->stackedWidget->setCurrentWidget(m_reportsWidget);
    statusBar()->showMessage("–†–µ–∂–∏–º: –û—Ç—á–µ—Ç—ã –∏ –∞–Ω–∞–ª–∏—Ç–∏–∫–∞");
}

void MainWindow::onRequestsButtonClicked()
{
    ui->stackedWidget->setCurrentWidget(m_requestsWidget);
    statusBar()->showMessage("–†–µ–∂–∏–º: –£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ –∑–∞—è–≤–∫–∞–º–∏");
}

void MainWindow::onUsersButtonClicked()
{
    RegistrationDialog dialog(this);
    dialog.setWindowTitle("–£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è–º–∏");
    
    if (dialog.exec() == QDialog::Accepted) {
        statusBar()->showMessage("–ü–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª–∏ –æ–±–Ω–æ–≤–ª–µ–Ω—ã", 3000);
    }
}

void MainWindow::onSettingsButtonClicked()
{
    QMessageBox::information(this, "–ù–∞—Å—Ç—Ä–æ–π–∫–∏", 
        "–ù–∞—Å—Ç—Ä–æ–π–∫–∏ —Å–∏—Å—Ç–µ–º—ã –±—É–¥—É—Ç –¥–æ—Å—Ç—É–ø–Ω—ã –≤ —Å–ª–µ–¥—É—é—â–µ–π –≤–µ—Ä—Å–∏–∏.");
}

void MainWindow::onExitButtonClicked()
{
    if (QMessageBox::question(this, "–ü–æ–¥—Ç–≤–µ—Ä–∂–¥–µ–Ω–∏–µ", 
        "–í—ã –¥–µ–π—Å—Ç–≤–∏—Ç–µ–ª—å–Ω–æ —Ö–æ—Ç–∏—Ç–µ –≤—ã–π—Ç–∏ –∏–∑ —Å–∏—Å—Ç–µ–º—ã?",
        QMessageBox::Yes | QMessageBox::No) == QMessageBox::Yes) {
        close();
    }
}